beacon_node/network/src/message_handler.rs

```diff
@@ -151,58 +152,137 @@ impl<T: BeaconChainTypes + 'static> MessageHandler<T> {
+                        );
+                    }
+                    RPCResponse::BeaconBlockHeaders(response) => {
+                        if let Some(decoded_block_headers) = self.decode_block_headers(response) {
+                            self.sync.on_beacon_block_headers_response(
+                                peer_id,
+                                decoded_block_headers,
+                                &mut self.network_context,
+                            );
+                        } else {
+                            warn!(self.log, "Peer sent invalid block headers";"peer" => format!("{:?}", peer_id))
+                        }
+                    }
+                    RPCResponse::BeaconBlockBodies(response) => {
+                        if let Some(decoded_block_bodies) = self.decode_block_bodies(response) {
+                            self.sync.on_beacon_block_bodies_response(
+                                peer_id,
+                                decoded_block_bodies,
+                                &mut self.network_context,
+                            );
+                        } else {
+                            warn!(self.log, "Peer sent invalid block bodies";"peer" => format!("{:?}", peer_id))
+                        }
+                    }
+                    RPCResponse::BeaconChainState(_) => {
+                        // We do not implement this endpoint, it is not required and will only likely be
+                        // useful for light-client support in later phases.
+                        //
+                        // Theoretically, we shouldn't reach this code because we should never send a
+                        // beacon state RPC request.
+                        warn!(self.log, "BeaconChainState RPC call is not supported.");
+                    }
+                }
             }
-        };
+        }
+    }
+
+    /// Verifies and decodes the ssz-encoded block bodies received from peers.
+    fn decode_block_bodies(
+        &self,
+        bodies_response: BeaconBlockBodiesResponse,
+    ) -> Option<DecodedBeaconBlockBodiesResponse> {
+        //TODO: Implement faster block verification before decoding entirely
+        let simple_decoded_bodies =
+            EncodeableBeaconBlockBodiesResponse::from_ssz_bytes(&bodies_response.block_bodies);
+
+        //TODO: Potentially improve the types used here for SSZ encoding/decoding
+        if let Ok(simple_decoded_bodies) = simple_decoded_bodies {
+            Some(DecodedBeaconBlockBodiesResponse {
+                block_roots: bodies_response
+                    .block_roots
+                    .expect("Responses must have associated roots"),
+                block_bodies: simple_decoded_bodies.block_bodies,
+            })
+        } else {
+            None
+        }
+    }
+
+    /// Verifies and decodes the ssz-encoded block headers received from peers.
+    fn decode_block_headers(
```

Comment:
This is a little nicer, but still 2 types are used. Surely we can do better ssz magic.